### ISSUE TITLE
Rule proposal: Enforce setting Error message before throwing

### ISSUE THREAD

**Issue by pvdlg** [sindresorhus/eslint-plugin-unicorn#128]
Throwing error without a message create confusion for users as it makes it difficult to know why the error happened. The rule would enforce that a message is set in the `Error` when it's thrown.

### Fail
```js
throw new Error();

throw new TypeError();

throw new EvalError();
```

### Pass
```js
throw new Error('message');

const error = new TypeError();
error.message = 'message';
throw error;
```

This rule (or another one?) would probably also have to check that the first parameter of subclassed errors is passed to the constructor or set to the `message` property.

**Comment by sindresorhus** [comment]
Good idea! Would you be interested in trying to implement it?

**Comment by sindresorhus** [comment]
> This rule (or another one?) would probably also have to check that the first parameter of subclassed errors is passed to the constructor or set to the message property.

I'm not sure about this. Sounds a little bit too opinionated. Not all errors are meant for human consumption.

**Comment by pvdlg** [comment]
> Good idea! Would you be interested in trying to implement it?

I never implemented an ESLint but I can try. Not sure how soon I can do that though.

**Comment by sindresorhus** [comment]
If you decide to do it, this gist might be helpful: https://gist.github.com/sindresorhus/1656c46f23545deff8cc713649dcff26

**Comment by johnwu93** [comment]
I would like to work on this feature.

When I work with exceptions in Javascript, I never set the `message` field of an error message of an error variable after creating in. For example, the situation that @pvdlg mentioned:
```javascript
const error = new TypeError();
error.message = 'message';
throw error;
```


In fact, I would never imagine somebody preferring this method of writing a message than setting it as a parameter on the constructor, i.e. `new Error('message')`. Maybe, we do not have to include this feature.

I am planning to write a rule that the message parameter must not be empty when throwing a newly instantiated error object.

**Comment by sindresorhus** [comment]
> I would like to work on this feature.

Great! Go ahead :)

> Maybe, we do not have to include this feature.

I agree. It doesn't make much sense. You can leave that out.

**Comment by johnwu93** [comment]
I am implementing this right now and I notice that there could be an edge case where the user passes in an empty string literal as a throw message. For example, you can have

```` javascript
throw new Error('')
````

Should this case be invalid?

**Comment by johnwu93** [comment]
Also, referring to @pvdlg's test case, should I cover the case where you are throwing a reference to an object?

For example, should this be an error:

```` javascript
const err = new Error();
throw err
````

**Comment by sindresorhus** [comment]
> throw new Error('')

Yes, should be invalid.

> const err = new Error();
throw err

This too should be invalid.
